File: scripts/update_crix_master.py
#!/usr/bin/env python3
"""
Fetch the latest Upbit listing master JSON and update the local snapshot.

Intended to be run on the server every ~2 hours (e.g. via cron/systemd timer).
The script compares the remote payload with the existing file, reports any new
coin codes, and always keeps `crix_master.json` current for the frontend.
"""

from __future__ import annotations

import argparse
import json
import sys
from gzip import decompress
from pathlib import Path
from typing import Iterable, List, Sequence
from urllib.error import URLError, HTTPError
from urllib.request import Request, urlopen


CRIX_MASTER_URL = "https://crix-static.upbit.com/v2/crix_master"
DEFAULT_OUTPUT = Path(__file__).resolve().parent.parent / "crix_master.json"
USER_AGENT = "UpbitListingMonitor/1.0 (+https://upbit.com)"


def read_remote(url: str) -> Sequence[dict]:
    """Download the JSON payload from Upbit."""
    request = Request(
        url,
        headers={
            "User-Agent": USER_AGENT,
            "Accept": "application/json",
            "Accept-Encoding": "gzip, deflate",
        },
    )
    with urlopen(request, timeout=30) as response:
        body = response.read()
        if response.headers.get("Content-Encoding") == "gzip":
            body = decompress(body)
        return json.loads(body.decode("utf-8"))


def read_local(path: Path) -> Sequence[dict]:
    if not path.exists():
        return []
    with path.open("r", encoding="utf-8") as handle:
        return json.load(handle)


def extract_codes(records: Iterable[dict]) -> set:
    return {record.get("code") for record in records if record.get("code")}


def write_json(path: Path, payload: Sequence[dict]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    tmp_path = path.with_suffix(path.suffix + ".tmp")
    with tmp_path.open("w", encoding="utf-8") as handle:
        json.dump(payload, handle, ensure_ascii=False, indent=2)
        handle.write("\n")
    tmp_path.replace(path)


def sync_crix_master(output: Path) -> int:
    try:
        remote_payload = read_remote(CRIX_MASTER_URL)
    except HTTPError as error:
        print(f"[update-crix] HTTP error: {error.code} {error.reason}", file=sys.stderr)
        return 1
    except URLError as error:
        print(f"[update-crix] Network error: {error.reason}", file=sys.stderr)
        return 1
    except Exception as error:  # pragma: no cover - defensive
        print(f"[update-crix] Unexpected error: {error}", file=sys.stderr)
        return 1

    local_payload = read_local(output)
    remote_codes = extract_codes(remote_payload)
    local_codes = extract_codes(local_payload)

    new_codes = sorted(remote_codes - local_codes)
    removed_codes = sorted(local_codes - remote_codes)

    if new_codes:
        print(f"[update-crix] {len(new_codes)} new listing(s): {', '.join(new_codes)}")
    if removed_codes:
        print(f"[update-crix] {len(removed_codes)} listing(s) removed upstream: {', '.join(removed_codes)}")

    if remote_payload != local_payload:
        write_json(output, remote_payload)
        print(f"[update-crix] Snapshot updated: {output}")
    else:
        print("[update-crix] No changes detected.")

    return 0


def parse_args(argv: Sequence[str]) -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Update crix_master.json from Upbit.")
    parser.add_argument(
        "-o",
        "--output",
        default=str(DEFAULT_OUTPUT),
        help=f"File path to write the snapshot (default: {DEFAULT_OUTPUT})",
    )
    return parser.parse_args(argv)


def main(argv: Sequence[str] | None = None) -> int:
    args = parse_args(argv or sys.argv[1:])
    output_path = Path(args.output).expanduser().resolve()
    return sync_crix_master(output_path)


if __name__ == "__main__":  # pragma: no cover
    raise SystemExit(main())

File: scripts/updateCrixMaster.js
#!/usr/bin/env node
"use strict";

const fsPromises = require("fs").promises;
const path = require("path");
const https = require("https");
const zlib = require("zlib");

const CRIX_MASTER_URL = "https://crix-static.upbit.com/v2/crix_master";
const PROJECT_ROOT = path.resolve(__dirname, "..");
const DEFAULT_OUTPUT = path.join(PROJECT_ROOT, "crix_master.json");
const USER_AGENT = "UpbitListingMonitor/1.0 (+https://upbit.com)";

function requestBuffer(url, { headers = {}, timeout = 30000 } = {}) {
  return new Promise((resolve, reject) => {
    const requestUrl = new URL(url);
    const options = {
      protocol: requestUrl.protocol,
      hostname: requestUrl.hostname,
      port: requestUrl.port,
      path: `${requestUrl.pathname}${requestUrl.search}`,
      method: "GET",
      headers: {
        "User-Agent": USER_AGENT,
        Accept: "application/json",
        "Accept-Encoding": "gzip, deflate",
        ...headers,
      },
    };

    const req = https.request(options, (res) => {
      const statusCode = res.statusCode || 0;
      if (statusCode >= 400) {
        const error = new Error(`Request failed with status ${statusCode}`);
        error.name = "HTTPError";
        error.statusCode = statusCode;
        error.statusMessage = res.statusMessage;
        res.resume();
        reject(error);
        return;
      }

      const chunks = [];
      res.on("data", (chunk) => chunks.push(chunk));
      res.on("end", () => {
        const buffer = Buffer.concat(chunks);
        const encoding = res.headers["content-encoding"] || "";
        if (encoding.includes("gzip")) {
          zlib.gunzip(buffer, (error, result) => {
            if (error) {
              reject(error);
            } else {
              resolve(result);
            }
          });
        } else if (encoding.includes("deflate")) {
          zlib.inflate(buffer, (error, result) => {
            if (error) {
              reject(error);
            } else {
              resolve(result);
            }
          });
        } else {
          resolve(buffer);
        }
      });
    });

    req.setTimeout(timeout, () => {
      req.destroy(new Error("Request timed out"));
    });

    req.on("error", (err) => reject(err));
    req.end();
  });
}

async function readRemote(url) {
  const body = await requestBuffer(url);
  return JSON.parse(body.toString("utf-8"));
}

async function readLocal(filePath) {
  try {
    const data = await fsPromises.readFile(filePath, "utf-8");
    return JSON.parse(data);
  } catch (error) {
    if (error.code === "ENOENT") {
      return [];
    }
    throw error;
  }
}

function extractCodes(records) {
  return new Set(
    (records || [])
      .map((record) => (record && record.code ? record.code : null))
      .filter(Boolean)
  );
}

async function writeJson(filePath, payload) {
  const directory = path.dirname(filePath);
  await fsPromises.mkdir(directory, { recursive: true });
  const tempPath = `${filePath}.tmp`;
  await fsPromises.writeFile(tempPath, JSON.stringify(payload, null, 2) + "\n", "utf-8");
  await fsPromises.rename(tempPath, filePath);
}

async function syncCrixMaster(outputPath) {
  let remotePayload;
  try {
    remotePayload = await readRemote(CRIX_MASTER_URL);
  } catch (error) {
    const message =
      error.name === "HTTPError"
        ? `[update-crix] HTTP error: ${error.statusCode} ${error.statusMessage || ""}`.trim()
        : `[update-crix] Network error: ${error.message}`;
    console.error(message);
    return 1;
  }

  let localPayload;
  try {
    localPayload = await readLocal(outputPath);
  } catch (error) {
    console.error(`[update-crix] Failed to read local snapshot: ${error.message}`);
    localPayload = [];
  }

  const remoteCodes = extractCodes(remotePayload);
  const localCodes = extractCodes(localPayload);

  const newCodes = Array.from(remoteCodes).filter((code) => !localCodes.has(code)).sort();
  const removedCodes = Array.from(localCodes).filter((code) => !remoteCodes.has(code)).sort();

  if (newCodes.length > 0) {
    console.log(`[update-crix] ${newCodes.length} new listing(s): ${newCodes.join(", ")}`);
  }

  if (removedCodes.length > 0) {
    console.log(`[update-crix] ${removedCodes.length} listing(s) removed upstream: ${removedCodes.join(", ")}`);
  }

  if (JSON.stringify(remotePayload) !== JSON.stringify(localPayload)) {
    await writeJson(outputPath, remotePayload);
    console.log(`[update-crix] Snapshot updated: ${outputPath}`);
  } else {
    console.log("[update-crix] No changes detected.");
  }

  return 0;
}

function parseArgs(argv) {
  const args = { output: DEFAULT_OUTPUT };
  for (let index = 0; index < argv.length; index += 1) {
    const token = argv[index];
    if ((token === "-o" || token === "--output") && argv[index + 1]) {
      args.output = path.resolve(argv[index + 1]);
      index += 1;
    }
  }
  return args;
}

async function main(argv = process.argv.slice(2)) {
  const args = parseArgs(argv);
  const exitCode = await syncCrixMaster(path.resolve(args.output));
  return exitCode;
}

if (require.main === module) {
  main().then(
    (code) => process.exitCode = code,
    (error) => {
      console.error(`[update-crix] Unexpected error: ${error.message}`);
      process.exitCode = 1;
    }
  );
}

module.exports = {
  CRIX_MASTER_URL,
  DEFAULT_OUTPUT,
  USER_AGENT,
  syncCrixMaster,
  readRemote,
  requestBuffer,
};
